ddprof-ffi/src/exporter.rs

```diff
@@ -275,8 +258,135 @@ pub unsafe extern "C" fn profile_exporter_send(
+        write!(f, "The tags string was malformed.")
+    }
+}
+
+impl Error for ParseTagsError {}
+
+fn parse_tag_chunk(chunk: &str) -> Result<Tag, ParseTagsError> {
+    if let Some(first_colon_position) = chunk.find(':') {
+        if first_colon_position == 0 {
+            return Err(ParseTagsError {
+                message: "Tag cannot start with a colon.",
+            });
+        }
+        let name = &chunk[..first_colon_position];
+        let value = &chunk[(first_colon_position + 1)..];
+        Ok(Tag {
+            name: CharSlice::from(name),
+            value: CharSlice::from(value),
+        })
+    } else {
+        Err(ParseTagsError {
+            message: "Unable to find a colon in the tag.",
+        })
+    }
+}
+
+/// Parse a string of tags typically provided by environment variables
+/// The tags are expected to be either space or comma separated:
+///     "key1:value1,key2:value2"
+///     "key1:value1 key2:value2"
+/// Tag names and values are required and may not be empty.
+fn parse_tags(str: &str) -> Result<crate::Vec<Tag>, ParseTagsError> {
+    /* The choice between spaces and comma as the separator is made by splitting
+     * by both and seeing which way matches the number of colons in the string.
+     *
+     * This is also done in a way to avoid allocations.
+     */
+    let colon_count = str.matches(':').count();
+    let comma_separated = str.split(',').map(parse_tag_chunk).filter(Result::is_ok);
+    let space_separated = str.split(' ').map(parse_tag_chunk).filter(Result::is_ok);
+
+    let seq = if comma_separated.clone().count() == colon_count {
+        comma_separated
+    } else if space_separated.clone().count() == colon_count {
+        space_separated
+    } else {
+        return Err(ParseTagsError {
+            message: "Unable to identify the tag separator; is the tags string malformed?",
+        });
+    };
+
+    let vec: Vec<_> = seq.flat_map(Result::ok).collect();
+    Ok(crate::Vec::from(vec))
+}
+
+#[repr(C)]
+pub enum VecTagResult<'a> {
+    Ok(crate::Vec<Tag<'a>>),
+    Err(crate::Vec<u8>),
+}
+
+#[export_name = "ddprof_ffi_VecTagResult_drop"]
+pub extern "C" fn vec_tag_result_drop(result: VecTagResult) {
+    std::mem::drop(result)
+}
+
+#[export_name = "ddprof_ffi_Vec_tag_parse"]
+pub extern "C" fn vec_tag_parse(string: CharSlice) -> VecTagResult {
+    match string.try_into() {
+        Ok(str) => match parse_tags(str) {
+            Ok(vec) => VecTagResult::Ok(vec),
+            Err(err) => VecTagResult::Err(crate::Vec::from(&err as &dyn Error)),
+        },
+
+        Err(err) => VecTagResult::Err(crate::Vec::from(&err as &dyn Error)),
+    }
+}
+
+#[allow(clippy::ptr_arg)]
+#[export_name = "ddprof_ffi_Vec_tag_clone"]
+pub extern "C" fn vec_tag_clone<'a>(vec: &'a crate::Vec<Tag<'a>>) -> VecTagResult {
+    let mut clone = Vec::new();
+    for tag in vec.into_iter() {
+        clone.push(tag.to_owned())
     }
+    VecTagResult::Ok(crate::Vec::from(clone))
 }
```

Comment:
I'm somewhat curious about this one -- what's the use case for cloning the vec?

Reply:
In PHP I store the result of parsing `DD_TAGS` globally. On each request, I clone these tags and add potentially specific values like runtime-id to the clone, which might help to support `fork` (it's not complete for PHP though, so maybe misguided).

Reply:
Hmmm... have you considered doing that by parsing `DD_TAGS` globally, adding them once at exporter creation, and then starting from an empty vec when adding the extra tags during `send`?

(Although I don't mind if you prefer doing it this way, seems harmless as well)